### PR TITLE
New design of cross-msg primitives (postbox and gas distribution)

### PR DESCRIPTION
**Pull request by cryptoAtwill**
Closes https://github.com/consensus-shipyard/ipc-actors/issues/18

## Changes

Adding `postbox` and `propagate` methods to simplify existing model.

- Introduced `postbox` related operation in `State.rs`
- Added `propagate` method in `lib.rs`
- Updated `apply_msg` to store message to `postbox` when current subnet is not the target subnet

But there are some issues/todos pending resolution, such as
- [x] Confirm highlevel flow is correct
- [x] Potential DDOS issue in nonce design
- [x] Potential simplification of postbox data structure as compared to that mentioned in #18
- [x] How to set gas properly and how to charge the propagation fee?

20230109 updates:
- Changed `postbox` owner to `Option<Vec<Address>>`
- Introduced gas constant
- Simplified `postbox` data structure to just HAMT of cid.

20230111 updates:
- Removed gas from `PostboxItem`
- Updated `postbox` hamt to HAMT<Cid, Vec<u8>>
- Added `whitelist_propagator` method
- Removed `Unknown` cross message type

## Tests

Tests are not written at this stage yet, will start updating the tests once the above issues are confirmed and resolved.

## Issues

- #18

### CODE REVIEW

**Comment by adlrocha** [comment]
# Updated design for nonces
To avoid potential vulnerabilities, censorship or congestion issues with nonces in the execution of cross-net messages with the introduction of the `postbox`, I suggest the following: 
- Every time a new top-down or bottom-up message is committed, its nonce is updated to the next subsequent one to determine the casual order against other cross-net messages. This nonce is used in the next subnet to enforce the execution of messages in order.  https://github.com/consensus-shipyard/ipc-actors/blob/4b7f17b4887b0f630d7ee317103fb52ca7254900/gateway/src/state.rs#L416-L418
- The `postbox` introduces a new stage in-between for the propagation of messages through several levels in the hierarchy, but even in this case, messages need to flow and be executed step-by-step through every intermediate subnet, which means that every time a message is propagated and conveniently committed in an intermediate subnet, the nonce is updated to the corresponding for the latest subnet (see https://github.com/consensus-shipyard/ipc-actors/blob/4b7f17b4887b0f630d7ee317103fb52ca7254900/gateway/src/lib.rs#L800) ensuring the casual order of propagation.

In short, we can still enforce that nonces from messages are executed sequentially (even more so now that we know that message need to necessarily flow through every intermediate subnet).

File: gateway/src/lib.rs [hunk 1]
```diff
@@ -683,43 +676,150 @@ impl Actor {
...
+
+    fn propagate<BS, RT>(rt: &mut RT, params: PropagateParams) -> Result<RawBytes, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        // TODO: update this to EOA
+        rt.validate_immediate_caller_accept_any()?;
+
+        let PropagateParams { gas, postbox_cid } = params;
```
Comment (adlrocha) [inline gateway/src/lib.rs:738]: We can remove `gas` from `PropagateParams`. As discussed sync, this will be a constant set by the actor. What we should check here is if the `rt.message().amount` is over the gas set for cross-messages. This will increase the internal balance of the actor.

We should also keep track of the amount of fees collected in a variable, as this will be used in the future to distribute rewards.

File: gateway/src/lib.rs [hunk 2]
```diff
@@ -683,43 +676,150 @@ impl Actor {
...
+            let postbox_item = st.load_from_postbox(rt.store(), postbox_cid).map_err(|e| {
+                log::error!("encountered error loading from postbox: {:?}", e);
+                actor_error!(unhandled_message, "cannot load from postbox")
+            })?;
+
+            if let Some(owners) = postbox_item.owners && !owners.contains(&owner) {
+                return Err(actor_error!(illegal_state, "owner not match"));
+            }
+
+            if postbox_item.gas < gas {
```
Comment (adlrocha) [inline gateway/src/lib.rs:751]: No need to perform this check, we should remove the `gas` field from cross-message items. This is determined by the actor.

File: gateway/src/lib.rs [hunk 3]
```diff
@@ -683,43 +676,150 @@ impl Actor {
...
+            Self::commit_cross_message(rt, st, cross_msg)?;
+            st.remove_from_postbox(rt.store(), postbox_cid)
+        })?;
+
+        Ok(RawBytes::default())
+    }
+
+    /// Commit the cross message to storage.
+    ///
+    /// ALWAYS CALL THIS IN A TRANSACTION.
```
Comment (adlrocha) [inline gateway/src/lib.rs:765]:
```suggestion
    /// NOTE: This function should always be called inside an `rt.transaction` 
```

File: gateway/src/lib.rs [hunk 4]
```diff
@@ -683,43 +676,150 @@ impl Actor {
...
+        let sto = cross_msg
+            .msg
+            .to
+            .subnet()
+            .map_err(|_| actor_error!(illegal_argument, "error getting subnet from msg"))?;
+        if sto == st.network_name {
+            return Err(actor_error!(illegal_state, "should already be committed"));
+        }
+
+        // TODO: runtime should expose is in transaction to ensure in transaction check
```
Comment (adlrocha) [inline gateway/src/lib.rs:784]: I don't understand this comment. Do you need any clarification or should we remove it?

File: gateway/src/state.rs [hunk 5]
```diff
@@ -481,6 +490,73 @@ impl State {
...
+        st: &BS,
+        owners: Option<Vec<Address>>,
+        gas: TokenAmount,
+        msg: CrossMsg,
+    ) -> anyhow::Result<()> {
+        let item = PostBoxItem::new(gas, msg, owners);
+        let cid: TCid<TLink<PostBoxItem>, PostBoxItem> = TCid::new_link(st, &item)?;
+        self.postbox.update(st, |postbox| {
+            let key = BytesKey::from(cid.cid().to_bytes());
+            postbox.set(key, ())?;
```
Comment (adlrocha) [inline gateway/src/state.rs:514]: Where are storing the message info?

File: gateway/src/state.rs [hunk 6]
```diff
@@ -481,6 +490,73 @@ impl State {
...
+    /// * `st` - The blockstore
+    /// * `owners` - The owners of the message that started the cross message. If None means
+    ///              anyone can propagate this message. Allows multiple owners.
+    /// * `gas` - The gas needed to propagate this message
+    /// * `msg` - The actual cross msg to store in `postbox`
+    pub(crate) fn insert_postbox<BS: Blockstore>(
+        &mut self,
+        st: &BS,
+        owners: Option<Vec<Address>>,
+        gas: TokenAmount,
```
Comment (adlrocha) [inline gateway/src/state.rs:507]: We can remove `gas` from the postbox item, as discussed this is set by the actor.

File: gateway/src/types.rs [hunk 7]
```diff
@@ -43,6 +48,40 @@ pub struct ApplyMsgParams {
     pub cross_msg: CrossMsg,
 }
 
+#[derive(Serialize_tuple, Deserialize_tuple, Clone)]
+pub struct PropagateParams {
+    pub gas: TokenAmount,
```
Comment (adlrocha) [inline gateway/src/types.rs:53]: Remove `gas` from params, this is no longer needed.

**Review by adlrocha — CHANGES_REQUESTED** [review]
We are really close. Just a few things pending: 
- Check the new approach for nonces and the execution of messages to see that it makes sense (it shouldn't require any change, just a sanity-check in tests to see that it works).
- Let's add a new function to add new owners for the propagation of messages (I think this is something users will need and is easy to include in this PR). 
- A few comments (let's remove the `gas` for postbox mesasges, etc.).

File: gateway/src/state.rs (hunk 5 above)
Comment (cryptoAtwill) [reply gateway/src/state.rs:514]: `TCid::new_link` will call `st.store`, we just need to store the cid to postbox.

**Comment by adlrocha** [comment]
I added a correct validation for the cross-net methods. This may have broken the build, as it depends on code included in this PR: https://github.com/consensus-shipyard/fvm-utils/pull/12 (I tested it locally). 

It may also have broken some of the `expect_validate_caller` in actor tests, we can check those before merging the PR.

File: gateway/src/lib.rs (hunk 4 above)
Comment (cryptoAtwill) [reply gateway/src/lib.rs:787]: I was thinking of to programatically enforce this, instead of in comments.

Comment (cryptoAtwill) [reply gateway/src/lib.rs:787]: We can remove this comment for now

File: gateway/src/lib.rs [hunk 8]
```diff
@@ -683,43 +676,150 @@ impl Actor {
...
+        match cross_msg.msg.apply_type(&st.network_name).map_err(|e| {
+            e.downcast_default(
+                ExitCode::USR_ILLEGAL_STATE,
+                "cannot convert cross message type",
+            )
+        })? {
+            IPCMsgType::Unknown => Err(actor_error!(illegal_state, "unknown cross message type")),
```
[the remark this answers is not in the document]
Comment (cryptoAtwill) [reply gateway/src/lib.rs:794]: removed

**Review by adlrocha — APPROVED** [review]
